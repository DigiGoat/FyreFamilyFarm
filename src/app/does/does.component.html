<app-goats type="Does"></app-goats>
