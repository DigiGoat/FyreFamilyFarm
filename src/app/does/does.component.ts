import { Component } from '@angular/core';

@Component({
  selector: 'app-does',
  templateUrl: './does.component.html',
  styleUrls: ['./does.component.scss']
})
export class DoesComponent {

}
