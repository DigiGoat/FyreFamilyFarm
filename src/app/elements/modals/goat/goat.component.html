<div #modal class="modal fade" id="exampleModal" tabindex="-1" aria-labelledby="exampleModalLabel" aria-hidden="true">
  <div class="modal-dialog modal-dialog-centered modal-dialog-scrollable modal-lg">
    <div class="modal-content">
      <div class="modal-header color-scheme-tertiary">
        <h1 class="modal-title fs-5" id="staticBackdropLabel">{{searchParam}}</h1>
        <button type="button" class="btn-close" data-bs-dismiss="modal" aria-label="Close"></button>
      </div>
      <div class="modal-body color-scheme-quaternary card">
        <h5 class="card-title text-center">{{name}}</h5>
        <h6 class="card-subtitle mb-2 text-center">"{{nickname}}"</h6>
        <div class="card-header color-scheme-quaternary">
          <div #carousel id="goatImageCarousel" class="carousel slide">
            <div class="carousel-indicators">
              <button *ngFor="let image of images; index as i" type="button" data-bs-target="#goatImageCarousel" [class.active]="i === 0"
                [attr.aria-current]="i === 0" attr.aria-label="Slide {{i}}" [attr.data-bs-slide-to]="i"></button>
            </div>
            <div class="carousel-inner d-flex align-items-center color-scheme-secondary" style="max-height: 75vh;">
              <div *ngFor="let image of images; index as i" #placeholderParent class="carousel-item placeholder-wave"
                [class.active]="i === 0">
                <img #imageRef (load)="imageRef.classList.remove('placeholder'); placeholderParent.classList.remove('placeholder-wave');"
                  [ngSrc]="image.file" [alt]="image.description" width="640" height="400" class="placeholder"
                  style="aspect-ratio: 4/3; width: 100%;object-fit:contain;object-position: center;"
                  [priority]="i < 2 || i === images!.length-1">
              </div>
              <div *ngFor="let image of images; index as i" class="carousel-item placeholder-wave" class="placeholder"
                [class.active]="i === 0">
                <div style="aspect-ratio: 4/3; width: 100%;object-fit:contain;object-position: center;"></div>
              </div>
            </div>
            <button class="carousel-control-prev" type="button" data-bs-target="#goatImageCarousel" data-bs-slide="prev">
              <span class="carousel-control-prev-icon" aria-hidden="true"></span>
              <span class="visually-hidden">Previous</span>
            </button>
            <button class="carousel-control-next" type="button" data-bs-target="#goatImageCarousel" data-bs-slide="next">
              <span class="carousel-control-next-icon" aria-hidden="true"></span>
              <span class="visually-hidden">Next</span>
            </button>
          </div>
          <div class="card-text text-center container color-scheme-tertiary rounded-bottom">
            <div class="text-capitalize"><span class="fw-bolder">Born</span>: <span class="fw-lighter">{{born | date:'longDate'}}</span>
              <br>
              <span class="fw-bolder">ID</span>: <span class="fw-lighter">{{id}}</span>
              <br>
              <span class="fw-bolder">Tattoo</span>: <span *ngFor="let animalTattoo of animalTattoos">
                {{animalTattoo.tattooLocation.name}}:<span class="fw-lighter">{{animalTattoo.tattoo}}</span>
              </span>
              <br>
              <span class="fw-bolder">Markings</span>: <span class="fw-lighter">{{colorAndMarking?.toLowerCase()}}</span>
            </div>
            {{description}}
          </div>
        </div>
      </div>
      <div class="modal-footer color-scheme-tertiary">
        <button type="button" class="btn btn-secondary" data-bs-dismiss="modal">Close</button>
      </div>
    </div>
  </div>
</div>
