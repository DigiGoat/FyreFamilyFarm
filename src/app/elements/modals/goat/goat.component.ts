import { Component, Input, type ElementRef, type OnDestroy, type OnInit, ViewChild, type AfterViewInit, ViewChildren, afterNextRender, afterRender } from '@angular/core';
import type { Goat } from '../../../services/goat/goat.service';
import { ImageService, type ImageEntry } from '../../../services/image/image.service';
import { ActivatedRoute, Router } from '@angular/router';


@Component({
  selector: 'app-goat-modal',
  templateUrl: './goat.component.html',
  styleUrl: './goat.component.scss'
})
export class GoatModalComponent implements OnInit, OnDestroy, AfterViewInit {
  @Input({ alias: 'goat', required: true }) goat?: Partial<Goat>;
  @Input({ required: true }) searchParam!: string;

  constructor(public imageService: ImageService, public router: Router, public route: ActivatedRoute) { }

  name?: string;
  nickname?: string;
  description?: string;
  id?: string;
  born?: string | number;
  animalTattoos?: Goat['animalTattoo'];
  colorAndMarking?: string;
  images?: ImageEntry[];
  ngOnInit(): void {
    this.name = this.goat?.name;
    this.nickname = this.goat?.nickname;
    this.description = this.goat?.description;
    this.id = this.goat?.normalizeId;
    this.born = this.goat?.dateOfBirth;
    this.animalTattoos = this.goat?.animalTattoo;
    this.colorAndMarking = this.goat?.colorAndMarking;
    this.images = this.imageService.getImages([this.id, this.name, this.nickname]);
  }

  ngOnDestroy() {
    this.modal?.hide();
    this.modal?.dispose();
    this.carousel?.dispose();
  }

  @ViewChild('modal') modalElement!: ElementRef<HTMLDivElement>;
  @ViewChild('carousel') carouselElement!: ElementRef<HTMLDivElement>;
  @ViewChildren('imageRef') imagRefs?: ElementRef<HTMLImageElement>[];
  private modal?: bootstrap.Modal;
  private carousel?: bootstrap.Carousel;
  ngAfterViewInit() {
    this.modal = bootstrap?.Modal.getOrCreateInstance(this.modalElement.nativeElement);
    this.carousel = bootstrap?.Carousel.getOrCreateInstance(this.carouselElement.nativeElement);
    this.open();
    this.modalElement.nativeElement.addEventListener('hidden.bs.modal', () => {
      this.router.navigate(['../'], { relativeTo: this.route });
    });
    this.imagRefs?.forEach(image => {
      image.nativeElement.addEventListener('error', () => {
        console.warn(`[${this.nickname ?? this.name ?? this.id}]`, 'Failed to load image');
        image.nativeElement.src = this.imageService.NotFound.file;
      }, { once: true });
    });
  }

  open() {
    this.modal?.show();
  }
  close() {
    this.modal?.hide();
  }

}
