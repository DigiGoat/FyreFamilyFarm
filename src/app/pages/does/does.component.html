<app-goats [goats]="goatService.does" name="Does"></app-goats>
