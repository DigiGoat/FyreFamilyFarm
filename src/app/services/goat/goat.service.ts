import { Observable } from 'rxjs';
import { retry } from 'rxjs/operators';

import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';


@Injectable({
  providedIn: 'root'
})
export class GoatService {


  constructor(private http: HttpClient) { }
  private _does: Goat[] = [];
  public does = new Observable<Goat[]>(observer => {
    if (this._does.length) {
      console.debug('Loaded Does From Cache', this._does);
      observer.next(this._does);
    } else {
      this.http.get<Goat[]>('/assets/resources/does.json')
        .pipe(
          retry(3), // retry a failed request up to 3 times
      )
        .subscribe({
          next: data => {
            this._does = data;
            console.debug('Loaded Does From Server', data);
            observer.next(data);
          },
          error: err => {
            if (err.status === 0) {
              // A client-side or network error occurred. Handle it accordingly.
              console.warn('An error occurred:', err.error);
            } else {
              // The backend returned an unsuccessful response code.
              // The response body may contain clues as to what went wrong.
              console.warn(
                `Backend returned code ${err.status}, body was: `, err.error);
            }
            // Return an observable with a user-facing error message.
            observer.error(err);
          }
        });
    }
  });
}
//export type Goat = (OwnedGoats['result']['items'][number] & { nickname: string; description: string; awards: Awards['result']['items']; colorAndMarking: string; /*obtained?: string;*/ });
export type Goat = {
  nickname?: string;
  name: string;
  description?: string;
  normalizeId: string;
  dateOfBirth: string;
};
export const Goat = {
  nickname: 'Your Goats Farm Name',
  name: 'YOUR GOATS ADGA NAME',
  description: 'A quick description about your goat',
  normalizeId: 'PD12345',
};
