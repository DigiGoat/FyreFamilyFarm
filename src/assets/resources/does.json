[
  {
    "name": "Highland Ranch AD Sandls N Sox",
    "dateOfBirth": "2019-03-05T00:00:00",
    "normalizeId": "PD2013808",
    "id": 1806160,
    "sex": "Female",
    "damId": 1572539,
    "sireId": 1730006,
    "ownerAccount": {
      "displayName": "Angel Fyre"
    },
    "colorAndMarking": "Buckskin With Random White And Frosting, Blue Eyes",
    "animalTattoo": [
      {
        "tattoo": "MSM2",
        "tattooLocation": {
          "name": "RE"
        }
      },
      {
        "tattoo": "L18",
        "tattooLocation": {
          "name": "LE"
        }
      }
    ],
    "nickname": "Juju"
  },
  {
    "name": "Gillis Family Farm Almond Roca",
    "dateOfBirth": "2020-03-23T00:00:00",
    "normalizeId": "PD2096889",
    "id": 1889367,
    "sex": "Female",
    "damId": 1766571,
    "sireId": 1490949,
    "ownerAccount": {
      "displayName": "Angel Fyre"
    },
    "colorAndMarking": "Chocolate Buckskin With Minimal White",
    "animalTattoo": [
      {
        "tattoo": "GFF5",
        "tattooLocation": {
          "name": "RE"
        }
      },
      {
        "tattoo": "M8",
        "tattooLocation": {
          "name": "LE"
        }
      }
    ]
  },
  {
    "name": "Fyre Famfarm Sweet'n Salty",
    "dateOfBirth": "2022-03-23T13:54:52",
    "normalizeId": "PD2227837",
    "id": 2227837,
    "sex": "Female",
    "damId": 1889367,
    "sireId": 1929176,
    "ownerAccount": {
      "displayName": "Angel Fyre"
    },
    "colorAndMarking": "Buckskin With Random White",
    "animalTattoo": [
      {
        "tattoo": "FF06",
        "tattooLocation": {
          "name": "RE"
        }
      },
      {
        "tattoo": "P1",
        "tattooLocation": {
          "name": "LE"
        }
      }
    ]
  },
  {
    "name": "Leisure Time ZZ Mscongeniality",
    "dateOfBirth": "2022-03-13T07:31:57",
    "normalizeId": "PD2238727",
    "id": 2238727,
    "sex": "Female",
    "damId": 1730922,
    "sireId": 1644245,
    "ownerAccount": {
      "displayName": "Angel Fyre"
    },
    "colorAndMarking": "Heavily Broken Buckskin, Brown Eyes",
    "animalTattoo": [
      {
        "tattoo": "1LTD",
        "tattooLocation": {
          "name": "RE"
        }
      },
      {
        "tattoo": "P10",
        "tattooLocation": {
          "name": "LE"
        }
      }
    ]
  },
  {
    "name": "Fyre Famfarm Little Bit Spicey",
    "dateOfBirth": "2023-04-28T16:51:17",
    "normalizeId": "PD2301632",
    "id": 2301632,
    "sex": "Female",
    "damId": 2227837,
    "sireId": 2185941,
    "ownerAccount": {
      "displayName": "Angel Fyre"
    },
    "colorAndMarking": "Buckskin With White Overlay, White Poll",
    "animalTattoo": [
      {
        "tattoo": "FF06",
        "tattooLocation": {
          "name": "RE"
        }
      },
      {
        "tattoo": "R1",
        "tattooLocation": {
          "name": "LE"
        }
      }
    ]
  },
  {
    "name": "Fyre Famfarm Heart Likea Truck",
    "dateOfBirth": "2023-06-06T15:05:01",
    "normalizeId": "PD2324819",
    "id": 2324819,
    "sex": "Female",
    "damId": 1806160,
    "sireId": 2157990,
    "ownerAccount": {
      "displayName": "Angel Fyre"
    },
    "colorAndMarking": "Buckskin, White Poll, Frosted Ears, Minimal White, Blue Eyes",
    "animalTattoo": [
      {
        "tattoo": "FF06",
        "tattooLocation": {
          "name": "RE"
        }
      },
      {
        "tattoo": "R4",
        "tattooLocation": {
          "name": "LE"
        }
      }
    ]
  },
  {
    "name": "Fyre Famfarm Watrmln Moonshine",
    "dateOfBirth": "2023-06-06T15:20:16",
    "normalizeId": "PD2324832",
    "id": 2324832,
    "sex": "Female",
    "damId": 1806160,
    "sireId": 2157990,
    "ownerAccount": {
      "displayName": "Angel Fyre"
    },
    "colorAndMarking": "Buckskin With White Spot Left Flank, White Poll, Blue Eyes",
    "animalTattoo": [
      {
        "tattoo": "FF06",
        "tattooLocation": {
          "name": "RE"
        }
      },
      {
        "tattoo": "R5",
        "tattooLocation": {
          "name": "LE"
        }
      }
    ]
  },
  {
    "name": "Fyre Famfarm Cowboy Queen",
    "dateOfBirth": "2023-06-28T12:00:32",
    "normalizeId": "PD2332443",
    "id": 2332443,
    "sex": "Female",
    "damId": 1889367,
    "sireId": 1850541,
    "ownerAccount": {
      "displayName": "Angel Fyre"
    },
    "colorAndMarking": "Light Chocolate Buckskin With Minimal White, White Poll",
    "animalTattoo": [
      {
        "tattoo": "FF06",
        "tattooLocation": {
          "name": "RE"
        }
      },
      {
        "tattoo": "R10",
        "tattooLocation": {
          "name": "LE"
        }
      }
    ]
  },
  {
    "name": "Fyre Famfarm Rua Wildflowers",
    "dateOfBirth": "2024-03-06T17:02:41",
    "normalizeId": "PD2371136",
    "id": 2371136,
    "sex": "Female",
    "damId": 2238727,
    "sireId": 2342517,
    "ownerAccount": {
      "displayName": "Angel Fyre"
    },
    "colorAndMarking": "Heavely Broken Light Buckskin, Browneyes",
    "animalTattoo": [
      {
        "tattoo": "FF06",
        "tattooLocation": {
          "name": "RE"
        }
      },
      {
        "tattoo": "S4",
        "tattooLocation": {
          "name": "LE"
        }
      }
    ]
  },
  {
    "name": "Alder Grove RR Lickety-Split",
    "dateOfBirth": "2024-04-30T17:29:10",
    "normalizeId": "PL2391326",
    "id": 2391326,
    "sex": "Female",
    "damId": 2284044,
    "sireId": 2323872,
    "ownerAccount": {
      "displayName": "Angel Fyre"
    },
    "colorAndMarking": "Sundgau With White Hind End",
    "animalTattoo": [
      {
        "tattoo": "4K",
        "tattooLocation": {
          "name": "RT"
        }
      },
      {
        "tattoo": "S9",
        "tattooLocation": {
          "name": "LT"
        }
      }
    ]
  },
  {
    "name": "Alder Grove RR Busy Bee",
    "dateOfBirth": "2024-04-16T18:04:02",
    "normalizeId": "PL2391327",
    "id": 2391327,
    "sex": "Female",
    "damId": 2292839,
    "sireId": 2323872,
    "ownerAccount": {
      "displayName": "Angel Fyre"
    },
    "colorAndMarking": "Sundgau ",
    "animalTattoo": [
      {
        "tattoo": "4K",
        "tattooLocation": {
          "name": "RT"
        }
      },
      {
        "tattoo": "S8",
        "tattooLocation": {
          "name": "LT"
        }
      }
    ]
  }
]
