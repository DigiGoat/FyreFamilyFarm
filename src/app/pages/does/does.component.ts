import { Component } from '@angular/core';

import { GoatService } from '../../services/goat/goat.service';


@Component({
  selector: 'app-does',
  templateUrl: './does.component.html',
  styleUrls: ['./does.component.scss']
})
export class DoesComponent {
  constructor(public goatService: GoatService) { }
}
