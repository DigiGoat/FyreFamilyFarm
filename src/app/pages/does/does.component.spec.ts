import { Observable } from 'rxjs';
import { of } from 'rxjs/internal/observable/of';

import { ComponentFixture, TestBed } from '@angular/core/testing';

import { GoatService } from '../../services/goat/goat.service';
import { DoesComponent } from './does.component';


jest.mock('../../services/goat/goat.service');
const goatService = jest.mocked(GoatService);
describe('DoesComponent', () => {
  let component: DoesComponent;
  let fixture: ComponentFixture<DoesComponent>;

  beforeEach(async () => {
    goatService.mockClear();
    await TestBed.configureTestingModule({
      declarations: [DoesComponent],
      providers: [GoatService],
    })
      .compileComponents();
    fixture = TestBed.createComponent(DoesComponent);
    component = fixture.componentInstance;
    TestBed.inject(GoatService);

    goatService.mockImplementation(() => {
      return {
        does: of([])
      } as unknown as GoatService;

    });
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });
  it('should call GoatService once', () => {
    expect(goatService).toHaveBeenCalledTimes(1);
  });
  it('should specify a name of Does', () => {
    expect(component.name).toBe('Does');
  });
  it('should specify an Observable getter of does', () => {
    expect(component.getter).toBeInstanceOf(Observable);
  });
  it('should match the snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });
});
