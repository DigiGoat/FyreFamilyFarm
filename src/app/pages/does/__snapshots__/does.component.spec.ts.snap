// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DoesComponent should match the snapshot 1`] = `
<app-does
  getter={[Function Observable]}
  goatService={[Function Object]}
  name={[Function String]}
>
  <app-goats />
</app-does>
`;
